### PR TITLE
Make ChatDispatcher compatible with plugins that listen for AsyncPlayerChatEvent

### LINKED ISSUE #304: Make ChatDispatcher compatible with plugins that listen to AsyncPlayerChatEvent
This will likely involve firing an `AsyncPlayerChatEvent` before sending the message in `ChatDispatcher`, and checking if its cancelled.

### PR DESCRIPTION
**Pull request by KingOfSquares**
Closes #304 

This will fire an AsyncPlayerChatEvent for every message sent with ChatDispatcher. If that event is canceled, the message will not be sent. 

This will also work with any feature that reads:
-The player sending the message
-The message sent
-The recipients of the message

This was tested with: 
https://www.spigotmc.org/resources/mute-system-1-8-x-1-14-x.15570/
https://www.spigotmc.org/resources/mute-chat-1-8-1-12.27156/

Signed-off-by: KingOfSquares <simonmorland@gmail.com>

### CODE REVIEW

File: core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java [hunk 1]
```diff
@@ -238,6 +238,9 @@ public void sendReply(Match match, Audience audience, MatchPlayer sender, @Text
 
   @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
   public void onChat(AsyncPlayerChatEvent event) {
+    // Ignore test messages from method "send"
+    if (event.getMessage().contains("8af5859a-1d6e-4fbd-ae5f-4c63ff630fa4")) return;
```
Comment (Electroid) [inline core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java]: This is too hacky, you'll need to develop another way.

Comment (Electroid) [reply core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java]: You could keep a `Cache` of recently sent `AsyncPlayerChatEvent`s.

Comment (KingOfSquares) [reply core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java]: I did a version of that now. Thanks for feedback.

File: core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java [hunk 2]
```diff
@@ -301,6 +321,27 @@ public void send(
...
+          new AsyncPlayerChatEvent(
+              false,
+              sender.getBukkit(),
+              message,
+              match.getPlayers().stream()
+                  .filter(filter)
+                  .map(MatchPlayer::getBukkit)
+                  .collect(Collectors.toSet()));
+
+      messageCache.put(event, "boring message");
```
Comment (Electroid) [inline core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java:338]: Just use a `""` string and explain why it's empty.

File: core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java [hunk 3]
```diff
@@ -301,6 +321,27 @@ public void send(
...
+
+      match.callEvent(event);
+
+      if (event.isCancelled()) return;
```
Comment (Electroid) [inline core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java:342]: We need to respect any changes to the event, say if the message is changed or the viewers are changed. So more than just check if cancelled.

Comment (KingOfSquares) [reply core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java:342]: Of course. I totally forgot that haha. Thanks!

File: core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java [hunk 4]
```diff
@@ -302,6 +322,29 @@ public void send(
...
+      AsyncPlayerChatEvent event =
+          new AsyncPlayerChatEvent(false, sender.getBukkit(), message, recipients);
+
+      messageCache.put(event, ""); // The value in the Cache is not used for anything
+
+      match.callEvent(event);
+
+      if (!event.getMessage().equals(message)) message = event.getMessage();
+      if (!event.getRecipients().equals(recipients)) recipients = event.getRecipients();
+      if (!event.getFormat().equals(format)) format = event.getFormat();
```
Comment (applenick) [inline core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java:344]: Figured out the issue.

You’re replacing `format` with the format provided by `AsyncPlayerChatEvent`, which in return is following the String [format syntax](https://docs.oracle.com/javase/8/docs/api/java/util/Formatter.html#syntax).

I would suggest simply omitting this check, as `send()` is provided a specific format to PGM (team/global/admin) and I don’t believe we need to care about changes made for that. 

Tested this change locally and all works well.

Comment (KingOfSquares) [reply core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java:344]: Ah, great catch. I will remove the check. Thanks!

**Comment by Electroid** [comment]
See 7a55789c034d3de1e85c5626762a55071e57b02b

**Pull request closed without merging**